File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>

using namespace std;

int main(void) {
	cin.tie(NULL);
	ios_base::sync_with_stdio(false);

	int k, n, right{ 2147483647 };
	long long int left{ 1 };

	cin >> k >> n;

	int* arr = new int[k];
	for (int i = 0; i < k; i++) {
		cin >> *(arr + i);
	}

	while (left <= right) {
		long long int count = 0;
		long long int mid = (left + right) / 2;

		for (int i = 0; i < k; i++) {
			count += *(arr + i) / mid;
		}

		if (count < n) {
			right = mid - 1;
		} else {
			left = mid + 1;
		}
	}

	cout << right << endl;

	delete[] arr;
	return 0;
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.util.Scanner;

public class Main {
    static int n, k, arr[];

    public static void main(String[] args) {
        input();
        solve();
    }

    public static void solve(){
        long left = 1;
        long right = 2147483647;

        while(left <= right){
            long mid = (left + right) / 2;

            long count = getCount(mid);
            if(count < n){
                right = mid - 1;
            } else{
                left = mid + 1;
            }
        }

        System.out.println(right);
    }

    public static long getCount(long size) {
        long count = 0;

        for(int len : arr){
            count += len / size;
        }

        return count;
    }

    public static void input(){
        Scanner sc = new Scanner(System.in);

        k = sc.nextInt();
        n = sc.nextInt();
        arr = new int[k];
        for(int i = 0; i < k; i++){
            arr[i] = sc.nextInt();
        }
    }
}
